feature – 6021 – Store original claim in Media object

vasconsaurus opened the pull request:
## Description

To debug items created using the API (e.g., through the Zapier integration), improve performance, and avoid duplications, it is important to store the `original_claim_url` in the `Media` object created when this attribute is present.

When a `ProjectMedia` is created (e.g., via a GraphQL mutation) and the `original_claim` attribute (a URL) is provided, a new `Media` instance is created from that URL and associated with the `ProjectMedia` instance. Currently, the reference between the original claim and the created media is lost. This reference needs to be stored.

References: CV2-6021

## How has this been tested?

### `project_media_7_test.rb`

"should save the original_claim url and original_claim_hash when Link Media is created from original_claim"
`rails test test/models/project_media_7_test.rb:200`

"should save the original_claim text and original_claim_hash when Claim media is created from original_claim"
`rails test test/models/project_media_7_test.rb:222`

"should save the original_claim url and original_claim_hash when UploadedAudio Media is created from original_claim"
`rails test test/models/project_media_7_test.rb:231`

"should check if the original_claim text exists and return that instance when trying to create claim media"
`rails test test/models/project_media_7_test.rb:247`

"should check if the original_claim url exists and return that instance when trying to create UploadedAudio media"
`rails test test/models/project_media_7_test.rb:256`

### `media_test.rb`

"Claim Media: should save the original_claim and original_claim_hash when created from original claim"
`rails test test/models/media_test.rb:631`

"Claim Media: should not save original_claim and original_claim_hash when not created from original claim"
`rails test test/models/media_test.rb:640`

"Claim Media: should not create duplicate media if media with original_claim_hash exists"
`rails test test/models/media_test.rb:646`

"Link Media: should save the original_claim and original_claim_hash when created from original claim"
`rails test test/models/media_test.rb:652`

"Link Media: should not save original_claim and original_claim_hash when not created from original claim"
`rails test test/models/media_test.rb:674`

"Link Media: should not create duplicate media if media with original_claim_hash exists"
`rails test test/models/media_test.rb:693`

"Uploaded Media: should save the original_claim and original_claim_hash when created from original claim"
`rails test test/models/media_test.rb:713`

"Uploaded Media: should not save original_claim and original_claim_hash when not created from original claim"
`rails test test/models/media_test.rb:729`

"Uploaded Media: should not create duplicate media if media with original_claim_hash exists"
`rails test test/models/media_test.rb:736`

## Things to pay attention to during code review

Please describe parts of the change that require extra attention during code review, for example:

- File FFFF, line LL: This refactoring does this and this. Is it consistent with how it’s implemented elsewhere?
- Etc.

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have added unit and feature tests, if the PR implements a new feature or otherwise would benefit from additional testing
- [ ] I have added regression tests, if the PR fixes a bug
- [ ] I have added logging, exception reporting, and custom tracing with any additional information required for debugging
- [ ] I considered secure coding practices when writing this code. Any security concerns are noted above.
- [ ] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [ ] My changes generate no new warnings
- [ ] If I added a third party module, I included a rationale for doing so and followed our current [guidelines](https://meedan.atlassian.net/wiki/spaces/ENG/overview#Choose-the-%E2%80%9Cright%E2%80%9D-3rd-party-module)

vasconsaurus commented:
@caiosba I moved the code from `ProjectMediaCreators` to `Media`, I think there is still work to be done, but I have some questions:

1. The code below was originally was in `ProjectMediaCreators`, so it was a concern. I think this makes more sense as a general concern instead of a Media method, but I'm not sure. Not sure where it would fit as a concern as well, maybe something related to http requests?
https://github.com/meedan/check-api/blob/384a84d569ca32ae2e0ecfb9e2adf793eee4cd40/app/models/media.rb#L129-L132

2. We moved creating claim, link and uploaded file medias to Media. But we still have a lot of very similar code still in `ProjectMediaCreators`. Does it make sense to keep both? Does it make sense to keep them in `Media` and `ProjectMediaCreators`?
https://github.com/meedan/check-api/blob/c115a52ab695f2f6062d394930f628b87deef6ab/app/models/concerns/project_media_creators.rb#L169-L206

3. Regarding `original_claim_hash`: right now every time I create a Media, I create a hash from the url or text. I'm wondering if there is a better way to do this. Maybe a callback that generates and add the hash when `original_claim` is present?

4. I'm wondering about the conceptual difference between a `ProjectMedia` that is created from an `original_claim` and a `ProjectMedia` that is created from a `Media` through the UI. Why are they handled differently? And, if a Media that wasn't created from an `original_claim` exists, then it won't have the `original_claim_hash`, then if we try to create it through zapier could we create a duplicate `Media`? I think I might be confusing `Media` and `ProjectMedia` again though.

caiosba commented:
@vasconsaurus , answers inline below:

> 1. The code below was originally was in `ProjectMediaCreators`, so it was a concern. I think this makes more sense as a general concern instead of a Media method, but I'm not sure. Not sure where it would fit as a concern as well, maybe something related to http requests?

I just saw that this method is used in only one place, do I don't think it should even be a method... I think you can just move the method body to the place where it's called.

> 2. We moved creating claim, link and uploaded file medias to Media. But we still have a lot of very similar code still in ProjectMediaCreators. Does it make sense to keep both? Does it make sense to keep them in Media and ProjectMediaCreators?

I see what you mean... let's leave those for later, in order to avoid confusion, and focus first on moving only the ones related to the `original_claim` workflow... it will be easier to compare after that.

> 3. Regarding original_claim_hash: right now every time I create a Media, I create a hash from the url or text. I'm wondering if there is a better way to do this. Maybe a callback that generates and add the hash when original_claim is present?

Sorry if I wasn't clear before! Yes, that should be a callback like `before_validate :generate_original_claim_hash, on: :create`. We have similar examples in the code base.

> 4. I'm wondering about the conceptual difference between a ProjectMedia that is created from an original_claim and a ProjectMedia that is created from a Media through the UI. Why are they handled differently? And, if a Media that wasn't created from an original_claim exists, then it won't have the original_claim_hash, then if we try to create it through zapier could we create a duplicate Media? I think I might be confusing Media and ProjectMedia again though.

That case will be harder to happen, because from the UI it's a file upload, which will probably have a different hash than the file hosted in a URL. Also, with the original claim hash approach, we can have a performance gain by avoiding an unnecessary HTTP request, while for a file upload, we need to receive the whole file anyway.

vasconsaurus commented:
@caiosba on my prior questions:
1. I moved the method body and removed the method
2. Moved the ones related to original_claim
3. Moved generating the original_claim_hash to a callback
4. Still want to talk about this in our 1:1

Besides that I also:
1. added uniqueness validation
2. added tests to Media
3. updated some tests in ProjectMedia

Do you think I can move this to ready for review? Or is there any work I forgot before doing that?

caiosba commented on app/models/media.rb:
```diff
@@ -103,4 +125,40 @@ def set_type
+
+  def self.create_uploaded_file_media_from_original_claim(media_type, url, ext)
+    klass = media_type.constantize
+    file = download_file(url, ext)
```
Please move this to inside the `else` block below - the idea of this method is to only download the file if the hash doesn't exist.

caiosba commented on app/models/media.rb:
```diff
@@ -103,4 +125,40 @@ def set_type
+    self.original_claim_hash = Digest::MD5.hexdigest(original_claim) unless self.original_claim.blank?
+  end
+
+  def self.create_uploaded_file_media_from_original_claim(media_type, url, ext)
```
I suggest you rename this method to `find_or_create_[...]`.

caiosba commented on app/models/media.rb:
```diff
@@ -103,4 +125,40 @@ def set_type
+    file
+  end
+
+  def self.create_claim_media_from_original_claim(text)
```
I suggest you rename this method to `find_or_create_[...]`.

caiosba commented on app/models/media.rb:
```diff
@@ -103,4 +125,40 @@ def set_type
+    Claim.find_by(original_claim_hash: Digest::MD5.hexdigest(text)) || Claim.create!(quote: text, original_claim: text)
+  end
+
+  def self.create_link_media_from_original_claim(url, project_media_team)
```
I suggest you rename this method to `find_or_create_[...]`.

caiosba commented on db/migrate/20250124155814_add_original_claim_to_media.rb:
```diff
@@ -0,0 +1,7 @@
+class AddOriginalClaimToMedia < ActiveRecord::Migration[6.1]
+  def change
+    add_column :medias, :original_claim, :text, null: true
+	  add_column :medias, :original_claim_hash, :string, null: true
+	  add_index :medias, :original_claim_hash, unique: true
```
Please fix the indentation.

caiosba commented on test/models/project_media_7_test.rb:
```diff
@@ -196,4 +196,76 @@ def setup
   end
+
+  test "should save the original_claim url and original_claim_hash when Link Media is created from original_claim" do
+    setup_elasticsearch
```
I don't think we need ElasticSearch in this test, right? Please remove if possible (it makes tests slower).

caiosba reviewed (changes requested):
Great stuff @vasconsaurus ! I left some comments, but it's great overall!

caiosba reviewed (approved):
Thanks for addressing all comments from the previous review, Manu!

:ship:

jayjay-w commented:
I reviewed this although it has already been merged and all the changes look good to me!

vasconsaurus commented:
Thank you @jayjay-w!
